R/breslow.R

```diff
@@ -0,0 +1,155 @@
+#' `xgboost.surv::sgb_bhaz()`.
+#'
+#' @return a `matrix` (obs x times). Number of columns is equal to `eval_times`
+#' and number of rows is equal to the number of test observations (i.e. the
+#' length of the `lp_test` vector). Depending on the `type` argument, the matrix
+#' can have either survival probabilities or cumulative hazard estimates.
+#'
+#' @references
+#' `r format_bib("cox_1972", "lin_2007")`
+#'
+#' @examples
+#' task = tsk("rats")
+#' part = partition(task, ratio = 0.8)
+#'
+#' learner = lrn("surv.coxph")
+#' learner$train(task, part$train)
+#' p_train = learner$predict(task, part$train)
+#' p_test  = learner$predict(task, part$test)
+#'
+#' surv = breslow(times = task$times(part$train), status = task$status(part$train),
+#'                lp_train = p_train$lp, lp_test = p_test$lp)
+#' head(surv)
+#' @export
+breslow = function(times, status, lp_train, lp_test, eval_times = NULL, type = "surv") {
+  assert_numeric(times, null.ok = FALSE)
+  assert_numeric(status, null.ok = FALSE)
+  assert_numeric(lp_train, null.ok = FALSE)
+  assert_true(length(times) == length(status))
+  assert_true(length(times) == length(lp_train))
+  assert_numeric(lp_test, null.ok = FALSE)
+  assert_numeric(eval_times, null.ok = TRUE)
+  assert_character(type, fixed = c("surv", "cumhaz"), null.ok = FALSE)
+
+  base_chaz = .cbhaz_breslow(times = times, status = status, lp = lp_train,
+                             eval_times = eval_times)
+
+  cumhaz = exp(lp_test) %*% t(base_chaz)
+  rownames(cumhaz) = seq(nrow(cumhaz))
+
+  if (type == "surv") {
+    return(exp(-cumhaz))
+  } else {
+    return(cumhaz)
+  }
+}
+
+# Breslow's Cumulative Baseline Hazard helper function
+#
+# Computes the cumulative baseline hazard using Breslow's (1972) estimator.
+# Assumes risk predictions are derived from a **proportional hazards**
+# regression model.
+#
+# @param times (`numeric()`)\cr Vector of times (train set).
+# @param status (`numeric()`)\cr Vector of status indicators (train set).
+# For each observation in the train set, this should be 0 (alive/censored) or
+# 1 (dead).
+# @param lp (`numeric()`)\cr Vector of linear predictors (train set).
+# These are the relative score predictions from a proportional hazards model
+# on the train set.
+# @param eval_times (`numeric()`)\cr Vector of times to compute the cumulative
+# baseline hazard estimations. If `NULL` (default), the unique and sorted
+# `times` from the train set will be used, otherwise the unique and sorted
+# `eval_times`.
+#
+# @returns
+# A vector of cumulative baseline hazards, one per (increasing) time
+# point. The times are added as names in the output vector.
+.cbhaz_breslow = function(times, status, lp, eval_times = NULL) {
+  # unique, sorted event times
+  event_times = sort(unique(times[status == 1]))
+
+  # baseline hazards are first evaluated on the `event_times`
+  bhaz = vapply(event_times, function(et) {
+    sum(times[status == 1] == et) / sum(exp(lp[times >= et]))
+  }, numeric(1))
```

Comment:
I'd recommend either wrapping this all in `cumsum` or make clear in the comment above this is not the cumulative hazard

Reply:
Either way is okay with me, I added a more informative comment.